### ISSUE TITLE
将 Keras 训练的模型转换成 iOS CoreML 可以使用的格式

### ISSUE THREAD

**Issue by Kamazuki** [Kamazuki/blogs#2]
### 前文提要：
#1 机器学习的开始 

### 本文目标

想要使用 iOS CoreML 的能力，得先把训练模型转换成 iOS CoreML 可以识别的格式

### 一、资料准备

1. Apple 的 coremltool 的 github 地址：https://github.com/apple/coremltools
2. Apple 的 CoreML 文档：https://developer.apple.com/documentation/coreml?language=objc
3. StackOverFlow 😂

### 二、环境准备：

1.  一台Mac

#### 1. 配置 python 虚拟环境，安装 coremltools https://github.com/apple/coremltools

    安装好并且激活了环境的命令行不需要关闭，后面继续在这个命令行里面工作。

#### 2. 创建一个目录，我这里创建了个 firstDemo 目录，位置可以自定义，然后命令行cd到这个目录，并且将之前训练好的模型放到这个目录。

    (pythonenv) ➜  firstDemo ls
    firstDemo.h5

本来我以为到这里就可以轻松进行模型转换了，查了下 coremltools 的文档才发现，这个 coremltools 只支持 keras 1.2.2 版本的模型，蛋碎了。因此需要修改一把keras环境，重新生成一个模型。

重新安装环境，第一步：

    #colab 顶部插入一行代码：
    !pip install tensorflow==1.14.0
    #执行成功之后，重启环境：菜单->代码执行程序->重新启动代码执行程序

第二步：

    #再插入一行代码：
    !pip install q keras==1.2.2
    #执行成功之后，重启环境：菜单->代码执行程序->重新启动代码执行程序

这样我们就将 keras 环境切换到了 1.2.2 版本

**_由于 keras 1.2.2 版本的 API 跟最新的有些许不同，所以需要改一下老的程序：_**

    import keras

    #Sequential : 模型的一种，采用顺序添加神经网络层的一种模型组织形式的类
    from keras.models import Sequential

    #Dense, Activation, Dropout : 神经网络的层，用于定义层的输入输出
    from keras.layers import Dense, Activation, Dropout

    #mnist : keras 提供的预定义数据
    from keras.datasets import mnist

    #RMSprop : keras预定义的一种优化器
    from keras.optimizers import RMSprop

**#========新代码========#**

    #这里需要import numpy.ndarray, numpy.asarray
    import numpy
    from numpy import ndarray, asarray

**#======^^^^^^^^^^^^======#**

    #这里有多个返回值
    (x_train, y_train), (x_test, y_test) = mnist.load_data()

    #原始加载出来的x_train\x_test 是 numpy.ndarray 的类型，由 28 * 28 的 uint8 矩阵组成？
    #需要把它们转化成一维的数据（后面定义的神经网络层只支持一维数据）
    x_train = x_train.reshape(60000, 28 * 28)
    x_test = x_test.reshape(10000, 28 * 28)

    #默认数据是 uint8 类型[0, 255]，后面要归一化，所以先转化成 float32 类型
    x_train = x_train.astype('float32')
    x_test = x_test.astype('float32')

    #这里的 x_train 实际上是 numpy.ndarray 类型，重载了一些运算符，除以255 进行归一化 
    x_train /= 255
    x_test /= 255

    #y_train\y_test 是 numpy.ndarray 类型，由 uint8 组成
    #因为我们需要做的是分类器（将输入的图片分类成 0～9 个分类）
    #因此将数据从 uint8 转化成长度为10的数组，转化后 result[i] = (uint8)(y_train == i)
    
**#========新代码========#**

    # keras 1.2.2 版本没有 keras.utils 这个模块
    # y_train = keras.utils.to_categorical(y_train, 10)
    # y_test = keras.utils.to_categorical(y_test, 10)

    temp_y = []
    for yt in y_train:
      temp_array = []
      temp_array.append((int(yt == 0)))
      temp_array.append((int(yt == 1)))
      temp_array.append((int(yt == 2)))
      temp_array.append((int(yt == 3)))
      temp_array.append((int(yt == 4)))
      temp_array.append((int(yt == 5)))
      temp_array.append((int(yt == 6)))
      temp_array.append((int(yt == 7)))
      temp_array.append((int(yt == 8)))
      temp_array.append((int(yt == 9)))
      temp_y.append(temp_array)

    y_train = asarray(temp_y)

    temp_y = []
    for yt in y_test:
      temp_array = []
      temp_array.append((int(yt == 0)))
      temp_array.append((int(yt == 1)))
      temp_array.append((int(yt == 2)))
      temp_array.append((int(yt == 3)))
      temp_array.append((int(yt == 4)))
      temp_array.append((int(yt == 5)))
      temp_array.append((int(yt == 6)))
      temp_array.append((int(yt == 7)))
      temp_array.append((int(yt == 8)))
      temp_array.append((int(yt == 9)))
      temp_y.append(temp_array)

    y_test = asarray(temp_y)

**#======^^^^^^^^^^^^======#**

    model = Sequential()

    #28*28 == 784, 这里是定义一个接受784个输入的神经网络层（对应一张图片），512个输出
    model.add(Dense(512, input_shape=(784,)))

    #激活层
    model.add(Activation('relu'))

    #随机丢弃部分数据，避免过拟合
    model.add(Dropout(0.2))

    #两层神经网络层之间，会自动匹配输入，所以这里只需要定义512个输出
    model.add(Dense(512))

    #激活层
    model.add(Activation('relu'))

    #随机丢弃部分数据，避免过拟合
    model.add(Dropout(0.2))

    #这里是最后一层，定义了10个输出，对应10个分类
    model.add(Dense(10))

    #使用了softmax 的激活函数来进行归一化，输出的数据范围[0, 1], 正好对应每一个分类的概率
    model.add(Activation('softmax'))

    #keres 预定义的优化函数
    opt = RMSprop()

    #编译模型
    model.compile(loss='mse', optimizer=opt, metrics=['accuracy'])

**#========新代码========#**

    #进行模型训练，20次迭代
    #keras 1.2.2 epoch 的名字是 nb_epoch
    model.fit(x_train, y_train, batch_size=128, nb_epoch=20)

**#======^^^^^^^^^^^^======#**

    #根据测试集来进行打分
    score = model.evaluate(x_test, y_test)

    #输出最后模型的得分
    print('Total loss on Testing Set:', score[0])
    print('Accuracy of Testing Set', score[1])

重新训练生成了新模型之后，就可以继续进行模型转换了

#### 3. 进行模型转换

模型转换的代码

    import coremltools
    model = coremltools.converters.keras.convert("firstDemo.h5", image_input_names='image')
    model.save("firstDemo.mlmodel")

保存到 firstDemo/convert.py

在刚刚还没有关闭的命令行工具里面，运行这个python文件

    python convert.py

接下来就能看到生成的 CoreML 模型文件啦

    (pythonenv) ➜  firstDemo ls
    convert.py        firstDemo.h5      firstDemo.mlmodel

### What's next？

接下来让我们将生成好的 CoreML 模型文件放到 iPhone 上面运行吧。
